sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExpressionCodegen.scala

```diff
@@ -0,0 +1,237 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions.codegen
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.expressions._
+
+/**
+ * Defines APIs used in expression code generation.
+ */
+object ExpressionCodegen {
+
+  /**
+   * Given an expression, returns the all necessary parameters to evaluate it, so the generated
+   * code of this expression can be split in a function.
+   * The 1st string in returned tuple is the parameter strings used to call the function.
+   * The 2nd string in returned tuple is the parameter strings used to declare the function.
+   *
+   * Returns `None` if it can't produce valid parameters.
+   *
+   * Params to include:
+   * 1. Evaluated columns referred by this, children or deferred expressions.
+   * 2. Rows referred by this, children or deferred expressions.
+   * 3. Eliminated subexpressions referred bu children expressions.
+   */
+  def getExpressionInputParams(
+      ctx: CodegenContext,
+      expr: Expression): Option[(Seq[String], Seq[String])] = {
+    val (inputAttrs, inputVars) = getInputVarsForChildren(ctx, expr)
+    val paramsFromColumns = prepareFunctionParams(ctx, inputAttrs, inputVars)
+
+    val subExprs = getSubExprInChildren(ctx, expr)
+    val subExprCodes = getSubExprCodes(ctx, subExprs)
+    val paramsFromSubExprs = prepareFunctionParams(ctx, subExprs, subExprCodes)
```

Comment:
do we support subexpression elimination when splitting codes to methods in the non-whole-stage-codegen path?

Reply:
I think so. The eliminated subexpressions will be extracted as parameters too.

Reply:
I mean before your PR.

Reply:
Oh, then it is no. Firstly we don't try to eliminate subexpressions before splitting codes now. Secondly, `splitExpressions` doesn't support subexpression elimination too.

Reply:
I think supporting subexpression elimination in `splitExpressions` really worth an individual JIRA, can we do it in a new PR? This PR should focus on supporting `splitExpressions` in whole stage codegen.

Reply:
If I get rid of the part of extracting subexpressions as parameters, some tests will be failed.

Because hash aggregation uses subexpression elimination under wholestage codegen, this PR enables splitting expression codes under wholestage codegen, so it inevitably needs to include subexpression parameters.

This PR doesn't yet to support `splitExpressions` in wholestage codegen. This is only applied to split codes of deeply nested expression now. But the added API can be applied to `splitExpressions` later. That is a TODO.

Reply:
Maybe we should move the supporting of subexpression elimination in `splitExpressions` to a new PR and merge that first. Reviewing 2 non-trivial optimizations at once is hard.

Reply:
This PR doesn't touch `splitExpressions`. It creates APIs that can be used to support subexpression elimination and wholestage codegen for `splitExpressions`. I plan to add those support in later PRs.

Actually there are not too much change related to subexpression elimination support in this PR. Only two APIs `getSubExprInChildren` and `getSubExprCodes` are specified for that. They are relatively small change.

If you still think it is better to split this part to a new PR that adds subexpression elimination support to `splitExpressions`, I can make it. What do you think?